In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from tqdm import tqdm

/home/user_3/anaconda3/lib/python3.8/site-packages/scipy/__init__.py:138: UserWarning: A NumPy version >=1.16.5 and <1.23.0 is required for this version of SciPy (detected version 1.23.1)
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion} is required for this version of "


In [2]:
df = pd.read_json('/fastcampus-data/log-data/log_sampling.json')

In [3]:
df

_metadata  \
0        {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1        {'bundled': ['Segment.io'], 'bundledIds': [], ...   
2        {'bundled': ['Segment.io'], 'bundledIds': [], ...   
3        {'bundled': ['Segment.io'], 'bundledIds': [], ...   
4        {'bundled': ['Segment.io'], 'bundledIds': [], ...   
...                                                    ...   
1019664  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1019665  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1019666  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1019667  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1019668  {'bundled': ['Segment.io'], 'bundledIds': [], ...   

                                  anonymousId channel  \
0        623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
1        623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
2        623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
3        b50480bc-1b18-449a-a073-39275cea3d82  client   
4        497f3995-1387-4c97-85af-183bcefac9c9  client   
...                                       ...     ...   
1019664  05480617-e770-4a51-86c6-6a1fdc01fc11  client   
1019665  3a3b5a94-d1ba-4838-91af-d578825bce4c  client   
1019666  3a3b5a94-d1ba-4838-91af-d578825bce4c  client   
1019667  7d0d05c7-c2f1-4a3d-b690-b07a8300f7a0  client   
1019668  3a3b5a94-d1ba-4838-91af-d578825bce4c  client   

                                                   context integrations  \
0        {'campaign': {}, 'ip': '220.124.80.98', 'libra...           {}   
1        {'ip': '220.124.80.98', 'library': {'name': 'a...           {}   
2        {'ip': '220.124.80.98', 'library': {'name': 'a...           {}   
3        {'campaign': {}, 'ip': '110.70.46.194', 'libra...           {}   
4        {'campaign': {}, 'ip': '104.28.102.64', 'libra...           {}   
...                                                    ...          ...   
1019664  {'ip': '58.122.5.19', 'library': {'name': 'ana...           {}   
1019665  {'ip': '110.8.170.95', 'library': {'name': 'an...           {}   
1019666  {'ip': '110.8.170.95', 'library': {'name': 'an...           {}   
1019667  {'ip': '1.215.13.164', 'library': {'name': 'an...           {}   
1019668  {'ip': '110.8.170.95', 'library': {'name': 'an...           {}   

                                         messageId         originalTimestamp  \
0        ajs-next-c1583afa7d4c5bf091c633d8232f9da3  2022-05-13T06:28:06.805Z   
1        ajs-next-9b8d710113206ffe6539e3ecc587567e  2022-05-13T06:30:54.677Z   
2        ajs-next-23c3daabb8984bada04b504404bbf197  2022-05-13T06:31:26.179Z   
3        ajs-next-04e4e02a6a6444550572ec5e2612e8e5  2022-05-13T06:33:01.601Z   
4        ajs-next-4286d9e866e25b67b8451692a02f297a  2022-05-13T06:40:33.953Z   
...                                            ...                       ...   
1019664  ajs-next-382a1a6c005807d79852ec4758975cd7  2022-08-24T01:22:21.069Z   
1019665  ajs-next-233a73a3bbcf9d4779c2d6740a4cb998  2022-08-24T01:51:34.048Z   
1019666  ajs-next-f817c9606f757a622fb99991cbc2d027  2022-08-24T01:52:12.507Z   
1019667  ajs-next-a828a06511701a8470d72e36f416e406  2022-08-24T02:09:59.937Z   
1019668  ajs-next-7f613656a89906621a30ffce1a70f5f8  2022-08-24T02:21:08.460Z   

                      projectId                receivedAt  \
0        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:28:06.815Z   
1        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:30:54.367Z   
2        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:31:25.869Z   
3        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:33:02.015Z   
4        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:40:34.317Z   
...                         ...                       ...   
1019664  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:22:21.563Z   
1019665  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:51:34.098Z   
1019666  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:52:12.554Z   
1019667  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T02:10:00.170Z   
1019668  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T02:21:08.683Z   

                        

In [17]:
df.columns

Index(['_metadata', 'anonymousId', 'channel', 'context', 'integrations',
       'messageId', 'originalTimestamp', 'projectId', 'receivedAt', 'sentAt',
       'timestamp', 'traits', 'type', 'userId', 'version', 'properties',
       'category', 'name', 'event'],
      dtype='object')

# preprocessing 코드 작성

- drop해야될 컬럼에 대해 전처리합니다.
- 컬럼의 dict 타입 key를 분리하는 작업을 합니다.
- 컬럼 dick 안 key 들에 대해서 drop해야될 것 전처리합니다.

In [3]:
# drop columns 전처리
def drop_columns(delete_columns:list,df:pd.DataFrame) -> pd.DataFrame:
    # 회의 때 얘기한 컬럼 전처리
    # None or 필요 없는 컬럼
    df = df.drop(columns=delete_columns)
    return df

def dict_to_column(columns:list,df:pd.DataFrame) -> pd.DataFrame:
    for col in columns:
        key_set = dict_to_set(col,df)
        df = set_to_column(col,key_set,df)
    return df

def dict_to_set(column:str,df:pd.DataFrame) -> set:
    key_set = set()
    for i in tqdm(df[column]):
        if isinstance(i, dict):
            key_set |= set(i.keys())
        elif isinstance(i, list) & len(i)>0:
            key_set |= set(i[0].keys())
    return key_set

def set_to_column(column:str,key_set:set,df:pd.DataFrame) -> pd.DataFrame:
    for key in key_set:
        
        #중복인 경우 컬럼_중복컬럼으로 추가
        if key in df.columns:
            df[column+'_'+key] = df[column].apply(lambda x: x.get(key, None) if isinstance(x, dict) else (x[0].get(key, None) \
                                       if isinstance(x,list)& len(x)>0 else None))
        #중복이 아닌 경우
        else:
            df[key] = df[column].apply(lambda x: x.get(key, None) if isinstance(x, dict) else (x[0].get(key, None) \
                                       if isinstance(x,list)& len(x)>0 else None))

    df = df.drop(columns=[column])
    return df

In [4]:
columns = ['_metadata','context','traits','properties']
df = dict_to_column(columns, df)
delete_columns = ['category','name','properties','type','userId','version','properties_properties', \
                 'properties_version','properties_type','properties_category','properties_name']
df = drop_columns(delete_columns, df)



100%|██████████| 1019669/1019669 [00:00<00:00, 1395748.78it/s]


In [5]:
df

anonymousId channel integrations  \
0        623cef3e-6160-4ca8-ab57-f1046a946cb1  client           {}   
1        623cef3e-6160-4ca8-ab57-f1046a946cb1  client           {}   
2        623cef3e-6160-4ca8-ab57-f1046a946cb1  client           {}   
3        b50480bc-1b18-449a-a073-39275cea3d82  client           {}   
4        497f3995-1387-4c97-85af-183bcefac9c9  client           {}   
...                                       ...     ...          ...   
1019664  05480617-e770-4a51-86c6-6a1fdc01fc11  client           {}   
1019665  3a3b5a94-d1ba-4838-91af-d578825bce4c  client           {}   
1019666  3a3b5a94-d1ba-4838-91af-d578825bce4c  client           {}   
1019667  7d0d05c7-c2f1-4a3d-b690-b07a8300f7a0  client           {}   
1019668  3a3b5a94-d1ba-4838-91af-d578825bce4c  client           {}   

                                         messageId         originalTimestamp  \
0        ajs-next-c1583afa7d4c5bf091c633d8232f9da3  2022-05-13T06:28:06.805Z   
1        ajs-next-9b8d710113206ffe6539e3ecc587567e  2022-05-13T06:30:54.677Z   
2        ajs-next-23c3daabb8984bada04b504404bbf197  2022-05-13T06:31:26.179Z   
3        ajs-next-04e4e02a6a6444550572ec5e2612e8e5  2022-05-13T06:33:01.601Z   
4        ajs-next-4286d9e866e25b67b8451692a02f297a  2022-05-13T06:40:33.953Z   
...                                            ...                       ...   
1019664  ajs-next-382a1a6c005807d79852ec4758975cd7  2022-08-24T01:22:21.069Z   
1019665  ajs-next-233a73a3bbcf9d4779c2d6740a4cb998  2022-08-24T01:51:34.048Z   
1019666  ajs-next-f817c9606f757a622fb99991cbc2d027  2022-08-24T01:52:12.507Z   
1019667  ajs-next-a828a06511701a8470d72e36f416e406  2022-08-24T02:09:59.937Z   
1019668  ajs-next-7f613656a89906621a30ffce1a70f5f8  2022-08-24T02:21:08.460Z   

                      projectId                receivedAt  \
0        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:28:06.815Z   
1        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:30:54.367Z   
2        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:31:25.869Z   
3        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:33:02.015Z   
4        pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:40:34.317Z   
...                         ...                       ...   
1019664  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:22:21.563Z   
1019665  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:51:34.098Z   
1019666  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T01:52:12.554Z   
1019667  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T02:10:00.170Z   
1019668  pAzAoFnKNpLfHxRmZ2oDF4  2022-08-24T02:21:08.683Z   

                           sentAt               timestamp      type  ...  \
0        2022-05-13T06:28:06.805Z 2022-05-13 06:28:06.815  identify  ...   
1        2022-05-13T06:30:54.677Z 2022-05-13 06:30:54.367  identify  ...   
2        2022-05-13T06:31:26.179Z 2022-05-13 06:31:25.869      page  ...   
3        2022-05-13T06:33:01.601Z 2022-05-13 06:33:02.015      page  ...   
4        2022-05-13T06:40:33.953Z 2022-05-13 06:40:34.317  identify  ...   
...                           ...                     ...       ...  ...   
1019664  2022-08-24T01:22:21.217Z 2022-08-24 01:22:21.415  identify  ...   
1019665  2022-08-24T01:51:34.055Z 2022-08-24 01:51:34.091      page  ...   
1019666  2022-08-24T01:52:12.512Z 2022-08-24 01:52:12.549      page  ...   
1019667  2022-08-24T02:09:59.987Z 2022-08-24 02:10:00.120      page  ...   
1019668  2022-08-24T02:21:08.647Z 2022-08-24 02:21:08.496  identify  ...   

        propertiesproperties  propertiesversion propertiestype  \
0                       None               None           None   
1                       None               None           None   
2                       None               None           None   
3                       None               None           None   
4                       None               None           None   
...                      ...                ...            ...   
1019664                 None               None           None   
1019665                 None               None           

# 분석 시작

In [132]:
dict_to_set('_metadata')

100%|██████████| 1019669/1019669 [00:00<00:00, 1285646.41it/s]


{'bundled', 'bundledIds', 'unbundled'}

In [133]:
dict_to_set('context')

100%|██████████| 1019669/1019669 [00:07<00:00, 129150.70it/s]


{'campaign', 'ip', 'library', 'locale', 'page', 'userAgent'}

In [135]:
dict_to_set('traits')

100%|██████████| 1019669/1019669 [00:00<00:00, 1798999.78it/s]


{'customerId'}

In [22]:
'''
category : 일치
name     : 5369번 빼고 일치 기존 컬럼을 쓰자
path : 새로움 컬럼
properties : 전부 none 드롭 (5369번 빼고)
referrer : 새로운 컬럼
search : 새로운 컬럼
title : 새로운 컬럼
type : 전부 none 드롭 (5369번 빼고)
url : 새로운 컬럼
userId : 전부 none 드롭 (5369번 빼고)
version : 전부 none 드롭 (5369번 빼고)
'''

def key_set(columns:str):
    key_set = set()
    for i in tqdm(df[columns]):
        if isinstance(i, dict):
            key_set |= set(i.keys())
    return key_set

key_set('properties')

100%|██████████| 1019669/1019669 [00:00<00:00, 1357902.37it/s]


['name',
 'title',
 'properties',
 'url',
 'userId',
 'version',
 'referrer',
 'category',
 'type',
 'path',
 'search']

In [14]:
key_set = set()
for i in tqdm(df.context):
    if isinstance(i, dict):
        key_set |= set(i.keys())
    for key in key_set:
        df = df.assign(
            property_category = df.properties.apply(lambda x: x.get(key, None) if isinstance(x, dict) else None)
        )
    

# 컬럼과 dict 컬럼 중복 확인 함수

In [7]:
def duplicate_confirm_columns(column:str, key:str):
    duplicate_confirm_series = df[column].apply(lambda x: x.get(key, None) if isinstance(x, dict) else None)
    different_cnt = (duplicate_confirm_series.fillna(0) != df[key].fillna(0)).sum()
    different_row = duplicate_confirm_series[(duplicate_confirm_series.fillna(0) != df[key].fillna(0))]
    return print(f'key name : {key}\n' + 'different_cnt :', different_cnt,'\n'+'different_row :', different_row, '\n')

## properites 컬럼 key 확인하기

In [8]:
# keys = key_set('properties')
# properties 컬럼 분해 & 기존 컬럼과 차이 확인
keys = ['category','name','type','userId','version']
for key in keys:
    duplicate_confirm_columns('properties', key)

key name : category
different_cnt : 0 
different_row : Series([], Name: properties, dtype: object) 

key name : name
different_cnt : 1 
different_row : 5369    CheckoutOrderComplete
Name: properties, dtype: object 

key name : type
different_cnt : 1019668 
different_row : 0          None
1          None
2          None
3          None
4          None
           ... 
1019664    None
1019665    None
1019666    None
1019667    None
1019668    None
Name: properties, Length: 1019668, dtype: object 

key name : userId
different_cnt : 1019668 
different_row : 0          None
1          None
2          None
3          None
4          None
           ... 
1019664    None
1019665    None
1019666    None
1019667    None
1019668    None
Name: properties, Length: 1019668, dtype: object 

key name : version
different_cnt : 1019669 
different_row : 0          None
1          None
2          None
3          None
4          None
           ... 
1019664    None
1019665    None
1019666    None
1019667   

# properties 컬럼안 key : type, userId, version 확인하기

In [37]:
property_type = df.properties.apply(lambda x: x.get('type', None) if isinstance(x, dict) else None)
property_type.isna().sum(), property_type.dropna()


(1019668,
 5369    page
 Name: properties, dtype: object)

In [36]:
property_userId = df.properties.apply(lambda x: x.get('userId', None) if isinstance(x, dict) else None)
property_userId.isna().sum(), property_userId.dropna()

(1019668,
 5369    76c6132d47e75c1c12634d133adc4b9037f1da32477df5...
 Name: properties, dtype: object)

In [38]:
property_version = df.properties.apply(lambda x: x.get('version', None) if isinstance(x, dict) else None)
property_version.isna().sum(), property_userId.dropna()

(1019668,
 5369    76c6132d47e75c1c12634d133adc4b9037f1da32477df5...
 Name: properties, dtype: object)

5369번 빼고 전부 nan

# properties 컬럼안 key: properties 확인하기

- 5369번 row 만 있고 다른건 다 na
- test 용도인 것으로  추정
- 전체 데이터에 대해서 학습할 시 테스트 용도 전처리 유의할 것
- 확인 방법으로 properties 컬럼안 key 중 name key에 대해서 na 값이 있으면 전처리하자


In [40]:
property_properites = df.properties.apply(lambda x: x.get('properties', None) if isinstance(x, dict) else None)
property_properites, property_properites.dropna()

(0          None
 1          None
 2          None
 3          None
 4          None
            ... 
 1019664    None
 1019665    None
 1019666    None
 1019667    None
 1019668    None
 Name: properties, Length: 1019669, dtype: object,
 5369    {'name': 'CheckoutOrderComplete', 'path': '/ch...
 Name: properties, dtype: object)

In [89]:
pd.DataFrame(df.iloc[5369,:]).T

_metadata  \
5369  {'bundled': ['Segment.io'], 'bundledIds': [], ...   

                               anonymousId channel  \
5369  3c8c87ee-ef2c-470c-8e11-3efe7f07b54a  client   

                                                context integrations  \
5369  {'ip': '121.152.176.140', 'library': {'name': ...           {}   

                                      messageId         originalTimestamp  \
5369  ajs-next-6366a0e975b9cf851887fa5a9c00b20e  2022-05-13T13:35:34.522Z   

                   projectId                receivedAt  \
5369  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T13:35:34.911Z   

                        sentAt               timestamp traits  type  \
5369  2022-05-13T13:35:34.522Z 2022-05-13 13:35:34.911   None  page   

                                                 userId version  \
5369  76c6132d47e75c1c12634d133adc4b9037f1da32477df5...       2   

                                             properties category  name event  
5369  {'name': 'CheckoutOrderComplete', 'path': '/',...     None  None  None

# properties 컬럼의 key 컬럼화

In [50]:
'''
path : 새로움 컬럼
referrer : 새로운 컬럼
search : 새로운 컬럼
title : 새로운 컬럼
url : 새로운 컬럼
위 properties key 들만 컬럼으로 추가
'''

keys = ['path','referrer','search','title','url']
for key in keys:
    df[key] = df.properties.apply(lambda x: x.get(key, None) if isinstance(x, dict) else None)

df = df.drop(columns=['property_category'])

### search None 전처리 후 데이터 확인하기

In [63]:
df['search'] = df['search'].apply(lambda x: None if x == {} else x)
search_series = df['search'].dropna()
search_series

### 궁금증
- search 가 none인 것들은 메인 페이지 혹은 메뉴 중 클릭하고 메인페이지에 해당하는 아이템을 클릭한 것일까??
- 비즈니스 분석 용도로는 어떤 검색을 많이하는 지 알수 있고 추천 시스템에 근거가 될 수도 있으므로 중요한 정보일 듯
- 총 100만건(1019669) 요청 중 86만건(863583)은 검색을 안함

## search 컬럼의 dict 따로 series 만들어 분석

- 너무많은 key들이 존재
- none이 많지만 분석여지가 충분

In [141]:
df.head()

_metadata  \
0  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
1  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
2  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
3  {'bundled': ['Segment.io'], 'bundledIds': [], ...   
4  {'bundled': ['Segment.io'], 'bundledIds': [], ...   

                            anonymousId channel  \
0  623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
1  623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
2  623cef3e-6160-4ca8-ab57-f1046a946cb1  client   
3  b50480bc-1b18-449a-a073-39275cea3d82  client   
4  497f3995-1387-4c97-85af-183bcefac9c9  client   

                                             context integrations  \
0  {'campaign': {}, 'ip': '220.124.80.98', 'libra...           {}   
1  {'ip': '220.124.80.98', 'library': {'name': 'a...           {}   
2  {'ip': '220.124.80.98', 'library': {'name': 'a...           {}   
3  {'campaign': {}, 'ip': '110.70.46.194', 'libra...           {}   
4  {'campaign': {}, 'ip': '104.28.102.64', 'libra...           {}   

                                   messageId         originalTimestamp  \
0  ajs-next-c1583afa7d4c5bf091c633d8232f9da3  2022-05-13T06:28:06.805Z   
1  ajs-next-9b8d710113206ffe6539e3ecc587567e  2022-05-13T06:30:54.677Z   
2  ajs-next-23c3daabb8984bada04b504404bbf197  2022-05-13T06:31:26.179Z   
3  ajs-next-04e4e02a6a6444550572ec5e2612e8e5  2022-05-13T06:33:01.601Z   
4  ajs-next-4286d9e866e25b67b8451692a02f297a  2022-05-13T06:40:33.953Z   

                projectId                receivedAt                    sentAt  \
0  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:28:06.815Z  2022-05-13T06:28:06.805Z   
1  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:30:54.367Z  2022-05-13T06:30:54.677Z   
2  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:31:25.869Z  2022-05-13T06:31:26.179Z   
3  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:33:02.015Z  2022-05-13T06:33:01.601Z   
4  pAzAoFnKNpLfHxRmZ2oDF4  2022-05-13T06:40:34.317Z  2022-05-13T06:40:33.953Z   

                timestamp                traits      type  \
0 2022-05-13 06:28:06.815  {'customerId': None}  identify   
1 2022-05-13 06:30:54.367  {'customerId': None}  identify   
2 2022-05-13 06:31:25.869                  None      page   
3 2022-05-13 06:33:02.015                  None      page   
4 2022-05-13 06:40:34.317  {'customerId': None}  identify   

                                              userId  version  \
0  3471de99bb6b2a801b0b5946389150999832d5132f2c51...        2   
1  3471de99bb6b2a801b0b5946389150999832d5132f2c51...        2   
2  3471de99bb6b2a801b0b5946389150999832d5132f2c51...        2   
3  7f84cdf255bd825cbe17ec63c7cb0822ff8e5cc49201dc...        2   
4  d8d3d56a1089728ddc8e489d5c9144c71d526464d1c756...        2   

                                          properties category  \
0                                               None     None   
1                                               None     None   
2  {'category': '', 'name': 'ArticlesArticle', 'p...            
3  {'path': '/', 'referrer': 'https://medistream....     None   
4                                               None     None   

              name event  
0             None  None  
1             None  None  
2  ArticlesArticle  None  
3             None  None  
4             None  None

In [98]:
key_set = set()
for key_dict in df.search:
    if isinstance(key_dict, dict):
        key_set |= set(key_dict.keys())
        
df_seach = pd.DataFrame()
for key in key_set:
        df_seach[key] = df.search.apply(lambda x: x.get(key, None) if isinstance(x, dict) else None)


In [102]:
df_seach.isnull().sum()

                   1019669
utm_id             1019666
utm_source          992661
c                  1019581
type               1013815
placeholder        1014605
boardId            1017182
boardTitle         1016771
articleId          1019387
shortlink          1019581
imp_uid            1019630
period             1015220
utm_term            992716
debugCustomerId    1019565
subOrder           1019618
paymentMethod      1019656
copyable           1019668
ids                1019145
referrer           1019623
nameType           1017152
af_dp              1019590
deep_link_value    1019581
utm_medium          992468
rank               1019564
imp_success        1019630
sort               1000711
limit              1008889
view               1017533
totalCount         1019618
keyword            1006580
pid                1019581
merchant_uid       1019630
category           1019660
utm_campaign        992664
napp               1019240
utm_content         993103
order              1019618
t

In [121]:
utm_columns = []
for col in df_seach.columns:
    if 'utm' in col:
        utm_columns.append(col)
        
utm_columns

['utm_id',
 'utm_source',
 'utm_term',
 'utm_medium',
 'utm_campaign',
 'utm_content']

In [118]:
for key in key_set:
    print(df_seach[key].dropna().value_counts(),'\n\n'+'-------------')
'''
'' : 지우자 다 none 이네
'utm_id' : 무슨컬럼이지.. 3개밖에 없음
'utm_source' : kakaotalk으로 27008개 카카오톡으로 결제?, 카카오톡 로그인?
'c' : 지우자
'type' : 중요한 정보일 것 같은데 none이 많아서.. 고민
'placeholder' : 메드스트림 홈페이지 내 어디서 검색했는 지 나옴
'boardId' : 
'''

## 컬럼 내용이

none이 너무 많아서 지워야 될게 많지만 분석해보고 싶습니다.
search 관련해서 제가 중점적으로 질문해도록 할게요
추천을 진행할 때 한의사분들은 검색을 잘 활용하지 않는 것 같아요
왜 검색을 안하는 지 그리고 검색한다면 어떤 것을 검색을 많이 하는 지

### search 로그 언제남나
- search 로그는 언제 남는 것인지 홈페이지 메인 페이지에서 검색해서 되는 것인지

### utm 이란
- utm 이 무엇인지??
- ['utm_id',
 'utm_source',
 'utm_term',
 'utm_medium',
 'utm_campaign',
 'utm_content']
- 특히, utm_term
- 메디스트림 알림 이뜨는 건가?, 공동구매 알림이 뜨는건가?, 댓글 알림이 뜨는건가?

### debugCustomerId
- userId가 안뜨면 찾아봤는 것인가?


Series([], Name: , dtype: int64) 

-------------
01GBMPVNV88JXFMAGP52S4P448    1
01GBW6JC74C6XQ3B0V4SN55WR5    1
01GC9NK7WHNANJNDNGZ8G367QH    1
Name: utm_id, dtype: int64 

-------------
kakaotalk    27008
Name: utm_source, dtype: int64 

-------------
ecp                85
appnotification     3
Name: c, dtype: int64 

-------------
article    4449
product     668
comment     664
user         43
wiki         30
Name: type, dtype: int64 

-------------
메디스트림 통합 검색         3162
일상 내 검색              802
채용·양도양수 내 검색         304
한의계 내 검색             190
아티클 내 검색             130
메디마켓 통합 검색           115
메디마켓 검색               79
게시판 내 검색              60
강의 내 검색               58
MEMBERS 내 검색          41
안드로메디 내 검색            35
임상 내 검색               14
공지사항 내 검색             13
유머 내 검색               10
경영 내 검색               10
경영세션 내 검색             10
예과생 내 검색               7
이벤트 내 검색               6
경영세션 EX 내 검색           5
부동산 내 검색               3
Hot 게시판 내 검색           3
<지금 우리 대학은> 내 검색  

realname                                                                                                                                 1561
nickname                                                                                                                                  954
realname](https://medistream.co.kr/users/cfab1ea7ae29aed50c84e2f61c508651aff794133723ee6af98e934fe3d168ec/articles?nameType=realname)       2
Name: nameType, dtype: int64 

-------------
medistream://    79
Name: af_dp, dtype: int64 

-------------
/account/shopping       85
/notificationsetting     3
Name: deep_link_value, dtype: int64 

-------------
KMDForest                                 11507
KMSForest                                  6929
AlarmForest                                2926
SeriesForest                               1289
ClinicContents                             1051
class                                       971
market                                      638
AndroMedi                

In [123]:
df.category.value_counts()

    464889
Name: category, dtype: int64

## intergrations 컬럼 분석

- 다 none 이여서 전처리 하자

In [126]:
df['integrations'] = df['integrations'].apply(lambda x: None if x =={} else x)

In [128]:
df['integrations'].value_counts()

Series([], Name: integrations, dtype: int64)